Sources/MapboxMaps/Location/Puck/Puck3D.swift

```diff
@@ -0,0 +1,98 @@
+        didSet {
+            guard isActive != oldValue else {
+                return
+            }
+            if isActive {
+                locationSource.add(self)
+                updateSourceAndLayer()
+            } else {
+                locationSource.remove(self)
+                if style.layerExists(withId: Self.layerID) {
+                    try! style.removeLayer(withId: Self.layerID)
+                }
+                if style.sourceExists(withId: Self.sourceID) {
+                    try! style.removeSource(withId: Self.sourceID)
+                }
+            }
+        }
+    }
+
+    internal var puckBearingSource: PuckBearingSource = .heading {
+        didSet {
+            updateSourceAndLayer()
+        }
+    }
+
+    private let configuration: Puck3DConfiguration
+    private let style: StyleProtocol
+    private let locationSource: LocationSourceProtocol
+
+    private static let sourceID = "puck-model-source"
+    private static let layerID = "puck-model-layer"
+
+    internal init(configuration: Puck3DConfiguration,
+                  style: StyleProtocol,
+                  locationSource: LocationSourceProtocol) {
+        self.configuration = configuration
+        self.style = style
+        self.locationSource = locationSource
+        super.init()
+    }
+
+    private func updateSourceAndLayer() {
+        guard isActive, let location = locationSource.latestLocation else {
+            return
+        }
+
+        var model = configuration.model
+        model.position = [location.coordinate.longitude, location.coordinate.latitude]
+        if model.orientation?.count != 3 {
+            if let invalidOrientation = model.orientation {
+                Log.warning(
+                    forMessage: "Puck3DConfiguration.model.orientation?.count must be 3 or nil. Actual orientation is \(invalidOrientation). Resetting it to [0, 0, 0].",
+                    category: "Puck")
+            }
+            model.orientation = [0, 0, 0]
+        }
+
+        switch puckBearingSource {
+        case .heading:
+            if let validHeadingDirection = location.headingDirection {
+                model.orientation?[2] += validHeadingDirection
+            }
+        case .course:
+            model.orientation?[2] += location.course
+        }
+
+        var source = ModelSource()
+        source.models = ["puck-model": model]
+
+        // update or create the source
+        if style.sourceExists(withId: Self.sourceID) {
+            try! style.setSourceProperties(for: Self.sourceID, properties: source.jsonObject())
+        } else {
+            try! style.addSource(source, id: Self.sourceID)
+        }
+
+        // create the layer if needed
+        if !style.layerExists(withId: Self.layerID) {
+            var modelLayer = ModelLayer(id: Self.layerID)
+            modelLayer.source = Self.sourceID
+            modelLayer.paint?.modelLayerType = .constant(.locationIndicator)
```

Comment:
Does it make sense to add assert here to guarantee for code reader that `modelLayer.paint` wouldn't be nil so we will assign variables correctly?
```suggestion
            assert(modelLayer.paint != nil)
            modelLayer.paint?.modelLayerType = .constant(.locationIndicator)
```